docs/create_your_own_brick.rst

```diff
@@ -0,0 +1,387 @@
+    ...         ...
+    ...     @apply.property('inputs')
+    ...     def apply_inputs(self):
+    ...         # Note that you can use any python code to define the name
+    ...         return self.fancy_name
+
+Using application calls
+"""""""""""""""""""""""
+
+If you need to add auxiliary variables to the computation graph
+:class:`.ComputationGraph`, you need to pass ``application_call`` as argument
+of your ``apply`` function, as shown in this example:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application
+    ...     def apply(self, x, application_call):
+    ...         application_call.add_auxiliary_variable(x.mean())
+    ...         return x + 1
+
+You can retrieve these annotated variables as usual with the computational
+graph :class:`.ComputationGraph` and filters :class:`.VariableFilter`.
+
+
+Lazy initialization
+~~~~~~~~~~~~~~~~~~~
+
+.. doctest::
+   :hide:
+
+    >>> from blocks.bricks import Feedforward, Linear
+    >>> from blocks.bricks.base import lazy, application
+    >>> from blocks.initialization import Constant
+
+Instead of forcing the user to provide all the brick attributes as arguments
+to the :meth:`.Brick.__init__` method, you could let him/her specify them
+later, after the creation of the brick. To enable this mechanism,
+called lazy initialization, you need to decorate the constructor with the 
+:func:`.lazy` decorator:
+
+    >>> @lazy(allocation=['attr1', 'attr2']) # doctest: +SKIP
+    ... def __init__(self, attr1, attr1)
+    ...     ...
+
+This allows the user to specify ``attr1`` and ``attr2`` after the creation of 
+the brick. For example, the following ``DoubleSequential`` brick is composed of
+two :class:`.Feedforward` bricks for which you do not need to
+specify the ``input_dim`` of ``brick2`` directly at its creation.
+
+    >>> class DoubleSequential(Feedforward):
+    ...     """Two sequential Feedforward bricks."""
+    ...     def __init__(self, brick1, brick2, **kwargs):
+    ...         super(Feedforward, self).__init__(**kwargs)
+    ...         self.brick1 = brick1
+    ...         self.brick2 = brick2
+    ...         self.input_dim = brick1.input_dim
+    ...         self.output_dim = brick2.output_dim
+    ...         self.children = [self.brick1, self.brick2]
+    ...
+    ...     def _push_allocation_config(self):
+    ...         self.brick2.input_dim = self.brick1.get_dim('output')
+    ...
+    ...     @application
+    ...     def apply(self, x):
+    ...         return self.brick2.apply(self.brick1.apply(x))
+
+Note how ``get_dim`` is used to retrieve the ``input_dim`` of ``brick1``. You
+can now use a ``DoubleSeuential`` brick as follows.
+
+    >>> brick1 = Linear(input_dim=3, output_dim=2, use_bias=False,
+    ...                 weights_init=Constant(2))
+    >>> brick2 = Linear(output_dim=4, use_bias=False, weights_init=Constant(2))
+    >>>
+    >>> seq = DoubleSequential(brick1, brick2)
+    >>> seq.initialize()
+    >>> brick2.input_dim
```

Comment:
Oh I see, you kind of have it in your next example, and you don't inherit from `Initializable` here. All right!